src/event/vrl_target.rs

```diff
@@ -0,0 +1,740 @@
+use super::{Event, EventMetadata, LogEvent, Metric, MetricKind, Value};
+use crate::config::log_schema;
+use lookup::LookupBuf;
+use snafu::Snafu;
+use std::{collections::BTreeMap, convert::TryFrom, iter::FromIterator};
+
+const VALID_METRIC_PATHS_SET: &str = ".name, .namespace, .timestamp, .kind, .tags";
+
+/// We can get the `type` of the metric in Remap, but can't set it.
+const VALID_METRIC_PATHS_GET: &str = ".name, .namespace, .timestamp, .kind, .tags, .type";
+
+/// Metrics aren't interested in paths that have a length longer than 3
+/// The longest path is 2, and we need to check that a third segment doesn't exist as we don't want
+/// fields such as `.tags.host.thing`.
+const MAX_METRIC_PATH_DEPTH: usize = 3;
+
+/// An adapter to turn `Event`s into `vrl::Target`s.
+#[derive(Debug, Clone)]
+pub enum VrlTarget {
```

Comment:
This is the bulk of this PR. It introduces an adapter for `Event` to satisfy `vrl::Target` to allow, for log events, `.` to be set to something other than a map (for example, here we want to allow it to be set to an array).